app/processing/optimize/algos.py

```diff
@@ -0,0 +1,352 @@
+        :return: Nodes that have no predecessor because of this.
+        """
+        result: list[ProcessedProgramNode] = []
+        succs = list(self.graph.successors(node.raw))
+        for succ in succs:
+            self.graph.remove_edge(node.raw, succ)
+            if not self.graph.pred[succ]:
+                result.append(self.graph.get_data_node(succ))
+        return result
+
+    def pop_nopred(self) -> ProcessedProgramNode:
+        """Remove and return the next nopred node.
+
+        This is a dummy implementation!
+
+        :return: Chosen node.
+        """
+        return self.nopred.pop()
+
+    def pop_uncomputable(self) -> ProcessedProgramNode:
+        """Remove and return the next node to be uncomputed.
+
+        Chooses the node with the fewest uncomputable qubits, i.e., the cheapest to uncompute.
+
+        :return: Chosen node.
+        """
+        if len(self.uncomputable) == 0:
+            msg = "No uncomputable nodes available to pop."
+            raise RuntimeError(msg)
+
+        result = min(
+            self.uncomputable,
+            key=lambda n: len(
+                n.qubit.returned_reusable_after_uncompute_ids,
+            ),
+        )
+
+        self.uncomputable.remove(result)
+        return result
+
+    def satisfy_dirty_qubit_requirement(self) -> None:
+        """Try to satisfy requirement for dirty qubits.
+
+        Add ancilla edges from previous nodes for this case.
+        Use the qubits in the order: dirty -> uncomputable -> reusable
+        """
+        if self.current_node is None:
+            raise RuntimeError
+
+        need_dirty = self.current_node.qubit.required_dirty_ids
+
+        while len(need_dirty) > 0 and len(self.dirty) > 0:
+            source = self.dirty[0]
+            possible_source_ids = source.qubit.returned_dirty_ids
+            size = min(len(need_dirty), len(possible_source_ids))
+            target_ids = need_dirty[:size]
+            need_dirty = need_dirty[size:]
+            source_ids = possible_source_ids[:size]
+            source.qubit.returned_dirty_ids = possible_source_ids[size:]
+            self.ancilla_edges.append(
+                AncillaConnection(
+                    (source.raw, source_ids),
+                    (self.current_node.raw, target_ids),
+                ),
+            )
+            if len(source.qubit.returned_dirty_ids) == 0:
+                self.dirty.remove(source)
+
+        while len(need_dirty) > 0 and len(self.uncomputable) > 0:
+            source = self.uncomputable[0]
+            possible_source_ids = source.qubit.returned_reusable_after_uncompute_ids
+            size = min(len(need_dirty), len(possible_source_ids))
+            target_ids = need_dirty[:size]
+            need_dirty = need_dirty[size:]
+            source_ids = possible_source_ids[:size]
+            source.qubit.returned_reusable_after_uncompute_ids = possible_source_ids[
+                size:
+            ]
```

Comment:
For me this looks better.
```suggestion
            source.qubit.returned_reusable_after_uncompute_ids = possible_source_ids[size:]
```